### DISCUSSION TITLE
Parameterize CLI script written in Deno

### DISCUSSION THREAD

**Discussion by quilicicf** [denoland/deno#8754] [Q&A]
Hi guys, 

I'm trying to do something that I'm not sure is supported in Deno and I am wondering if I missed a feature or if there is a possible improvement to Deno. 

## What

I'm trying to write a CLI tool that's supposed to have access to a single folder on the user's machine. The idea:

1. the whole configuration/assets live in a single folder (meaning, copy/pasting this folder replicates the whole behavior of the tool)
2. the tool can only read/write to that folder and no other
3. it asks for minimal permissions

So, for example, if a user wants to install my tool in `/home/me/forge`, I'd like them to be able to install it with `deno install --allow-read=/home/me/forge` AND I'd like my script to know where that folder is. 

## How

I have tried the options below:

* read the folders the tool is allowed to read at runtime => not possible (probably security reasons?)
* put the configuration (containing the folder path) in HOME => requires `allow-env`, breaks `#1`, two-step installation
* inject the project folder at installation => `deno install` currently can't do that, I'd have to write a wrapping installation script. Seems convoluted and does not inspire trust

From my perspective, this looks like a use-case others might run into and it would be nice to have it supported by Deno It would probably best be solved at installation (new capabilities in `deno install`?).

WDYT? Any solution I missed?

**Comment by harshagarwalnyu** [comment]
Careful, import.meta.url won't give you the install directory. When you deno install a script from a remote URL, the generated shim still points at that remote URL, so at runtime import.meta.url is something like https://.../mod.ts, not a local path. new URL(".", import.meta.url).pathname just gives you the remote path portion, and import.meta.filename / import.meta.dirname are only set for local file:// modules, so they're undefined for a remotely installed script.

For parameterizing the CLI, the runtime-native ways are:

1) Command-line args, simplest:

  const [configPath] = Deno.args;
  // deno run --allow-read mytool /path/to/config.json

2) An env var:

  const dir = Deno.env.get("MYTOOL_HOME") ?? "";

3) A fixed config location derived from the home dir, e.g. Deno.env.get("HOME") + "/.config/mytool/", and scope your --allow-read / --allow-write to exactly that.

Any of those is stable regardless of where the tool is installed. The import.meta.url route only works when you run the script directly from a local file, not through deno install.
